query/src/main/java/org/infinispan/query/impl/EntityLoader.java

```diff
@@ -23,20 +25,29 @@ public EntityLoader(AdvancedCache<?, ?> cache, KeyTransformationHandler keyTrans
       this.cache = cache;
    }
 
+   private Object decodeKey(EntityInfo entityInfo) {
+      return keyTransformationHandler.stringToKey(entityInfo.getId().toString(), cache.getClassLoader());
+   }
+
    public Object load(EntityInfo entityInfo) {
-      Object cacheKey = keyTransformationHandler.stringToKey(entityInfo.getId().toString(), cache.getClassLoader());
-      return cache.get(cacheKey);
+      return cache.get(decodeKey(entityInfo));
    }
 
-   public List<Object> load(Collection<EntityInfo> entityInfos) {
-      ArrayList<Object> list = new ArrayList<Object>(entityInfos.size());
+   public List<Object> load(List<EntityInfo> entityInfos) {
+      int entitiesSize = entityInfos.size();
+      Set<Object> keys = new LinkedHashSet<>(entitiesSize);
       for (EntityInfo e : entityInfos) {
-         Object entity = load(e);
-         if (entity != null) {
-            list.add(entity);
+         keys.add(decodeKey(e));
+      }
+      Map<?, ?> entries = cache.getAll(keys);
+      List<Object> results = new LinkedList<>();
+      for (Object key : keys) {
```

Comment:
Unfortunately he needs the results in the same order that the keys were provided in.

I talked with Gustavo about this very briefly.  Thinking about it more after, we could return an ordered map so he doesn't have to do all the lookups, but we would require having a returned map that had null values to keep the ordering.  This is doable, but right now our APIs are not defined that way.  Doing this could actually lead to some serialization overhead improvements for remote getAll as well though - since we wouldn't have to return the keys (I was thinking of reevaluating this for ISPN 8 with HotRod 3.0 maybe)